# Tranposable Elements

## Create a Repeat Database with Repeat Modeler

Based on Minal's statistics, I
1) excluded all heterozygotes F2, F5 and F6
2) Focused on hap1s because they seem to be the longest assemblies overall
3) From F assemblies, I chose the longest and more contiguous (F4 and F3)
4) From CS assemblies, I chose the longest (CS7) and the least fragmented (C4)

Run repeat Modeler for the chosen da
~~~bash
ml load bioinfo-tools RepeatModeler/2.0.1
ml load gnuparallel

for i in $(ls *fa); do BuildDatabase -name ${i/.fa/.DB} -engine rmblast $i ; done

RepeatModeler -database CS4.hap1.DB -pa 18
RepeatModeler -database CS7.hap1.DB -pa 18
RepeatModeler -database F3.hap1.DB -pa 18
RepeatModeler -database F4.hap1.DB -pa 18

cp ../RM_20972.TueOct261710022021/consensi.fa.classified ./CS4_consensi.fa.classified
cp ../RM_24115.TueOct261737382021/consensi.fa.classified ./F3_consensi.fa.classified
cp ../RM_27732.TueOct261748332021/consensi.fa.classified ./F4_consensi.fa.classified
cp ../RM_791.TueOct261726362021/consensi.fa.classified ./CS7_consensi.fa.classified

ml load perl
ml load bioinfo-tools BioPerl/1.7.2_Perl5.26.2

perl /proj/snic2020-2-19/private/herring/users/mafalda/software/CRL_Scripts1.0/repeatmodeler_parse.pl --fastafile CS4_consensi.fa.classified --unknowns CS4_consensi.fa.classified_unknowns.fasta --identities CS4_consensi.fa.classified_identities.fasta
for i in $(ls *_consensi.fa.classified); do perl /proj/snic2020-2-19/private/herring/users/mafalda/software/CRL_Scripts1.0/repeatmodeler_parse.pl --fastafile $i --unknowns "$i"_unknowns.fasta --identities "$i"_identities.fasta; done

~~~

Find the unknowns
~~~bash
ml load bioinfo-tools blast/2.11.0+
makeblastdb -in Tpases020812 -dbtype prot
ml load gnuparallel

parallel -j 4 'blastx -query {1} -db DB/Tpases020812  -evalue 1e-10 -num_descriptions 10 -out {1}_modelerunknown_blast_results.txt' ::: $(ls *.classified_unknowns.fasta)
~~~

~~~bash
samples="CS4 CS7 F3 F4"

for i in ${samples}; do

perl /proj/snic2020-2-19/private/herring/users/mafalda/software/CRL_Scripts1.0/transposon_blast_parse.pl --blastx ${i}_consensi.fa.classified_unknowns.fasta_modelerunknown_blast_results.txt --modelerunknown ${i}_consensi.fa.classified_unknowns.fasta

mv unknown_elements.txt ${i}_ModelerUnknown.lib

cat identified_elements.txt ${i}_consensi.fa.classified_identities.fasta  > ${i}_ModelerID.lib

done;
~~~

Blast against Clupea harengus genes:

~~~bash
ml load bioinfo-tools blast/2.11.0+
ml load gnuparallel

parallel -j 4 'blastx -query {1}_ModelerUnknown.lib -db DB/Clupea_harengus.Ch_v2.0.2.pep.all.fa -evalue 1e-10 -outfmt "6 qseqid staxids bitscore std sscinames sskingdoms stitle" -max_target_seqs 25 -culling_limit 2 -num_threads 18 -out {1}_ModelerUnknown.lib_blast_results.fmt6.txt' ::: $(echo "CS4 CS7 F3 F4")

parallel -j 4 'blastx -query {1}_ModelerID.lib -db DB/Clupea_harengus.Ch_v2.0.2.pep.all.fa -evalue 1e-10 -outfmt "6 qseqid staxids bitscore std sscinames sskingdoms stitle" -max_target_seqs 25 -culling_limit 2 -num_threads 18 -out {1}_ModelerID.lib_blast_results.fmt6.txt' ::: $(echo "CS4 CS7 F3 F4")
~~~

Remove the ones that map to genes:
~~~bash
filterbyname.sh in=CS4_ModelerUnknown.lib out=CS4_ModelerUnknown.NoProt.fasta names=$(cut -f1 CS4_ModelerUnknown.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=CS7_ModelerUnknown.lib out=CS7_ModelerUnknown.NoProt.fasta names=$(cut -f1 CS7_ModelerUnknown.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=F3_ModelerUnknown.lib out=F3_ModelerUnknown.NoProt.fasta names=$(cut -f1 F3_ModelerUnknown.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=F4_ModelerUnknown.lib out=F4_ModelerUnknown.NoProt.fasta names=$(cut -f1 F4_ModelerUnknown.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=CS4_ModelerID.lib out=CS4_ModelerID.NoProt.fasta names=$(cut -f1 CS4_ModelerID.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=CS7_ModelerID.lib out=CS7_ModelerID.NoProt.fasta names=$(cut -f1 CS7_ModelerID.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=F3_ModelerID.lib out=F3_ModelerID.NoProt.fasta names=$(cut -f1 F3_ModelerID.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
filterbyname.sh in=F4_ModelerID.lib out=F4_ModelerID.NoProt.fasta names=$(cut -f1 F4_ModelerID.lib_blast_results.fmt6.txt | sort | uniq | sed -z 's/\n/,/g' | sed 's/.$//') include=false overwrite=true
~~~

Combine the libraries

~~~bash
for i in $(ls *.NoProt.fasta); do name=${i/_*/}; sed 's/>/>'"$name"'_/g' $i > ${i/.fasta/.FixedHead.fasta} ; done

cat *NoProt.fasta > atlanticherring_combined_repDB_noprot.fasta
~~~

Run CDHist to remove redundancy in the library.

For CDHIT, I will use the same parameters as in this paper:
https://academic.oup.com/database/article/doi/10.1093/database/bax106/4812028#129646603

~~~bash
ml load bioinfo-tools cd-hit/4.8.1 

write the command down:

cd-hit-est -i atlanticherring_combined_repDB_noprot.fasta -o atlanticherring_combined_repDB_noprot_90.fasta -c 0.9 â€“n 8 -d 0 -M 1600 -T 8
~~~

## Finding inversion sequences in Hifiasm assemblies

Minal scaffolded BS3 and CS10 for each inversion. I used these scaffolded inversions to find the breakpoints of the inversions in the other assemblies and run RepeatMasker for each one of them.

~~~bash
ml load bioinfo-tools samtools/1.12

samtools faidx CS10.hap1.fasta chr6_inv:1144464-3774578 > CS10.hap1.chr6.fasta
samtools faidx CS10.hap1.fasta chr12_inv:1156916-8338930 > CS10.hap1.chr12.fasta
samtools faidx CS10.hap1.fasta chr17_inv:741635-2249708 > CS10.hap1.chr17.fasta
samtools faidx CS10.hap1.fasta chr23_inv:854555-2134906 > CS10.hap1.chr23.fasta

samtools faidx BS3.hap2.fasta chr6_inv:1051601-3747148 > BS3.hap2.chr6.fasta
samtools faidx BS3.hap2.fasta chr12_inv:1215813-8428837 > BS3.hap2.chr12.fasta
samtools faidx BS3.hap2.fasta chr17_inv:687774-2468994 > BS3.hap2.chr17.fasta
samtools faidx BS3.hap2.fasta chr23_inv:788123-1676584 > BS3.hap2.chr23.fasta
~~~

We need to change the name of the fasta inside the fasta file. Here, we split the name of the input file using an array. The array stores the characters separated by ".", and then we access the array's 3rd position (2, because bash is 0-based) to find the name of the chromosome.

~~~bash
for i in $(ls *chr*.fasta); do arrChr=(${i//./ }); sed -i "s/>.*/>${arrChr[2]}_inv/" $i; done
~~~

Now, let's extract the first part of the inversion. We use the same array trick to construct the faidx command.

~~~bash
# First 10kb
for i in $(ls *.chr*.fasta); do arrChr=(${i//./ }); samtools faidx $i ${arrChr[2]}"_inv:1-10000" > ${i/.fasta/.start.fasta}; done
~~~

Last 10kb, with the script `obtain_coordinates.sh`

~~~bash
for INPUT in $(ls *chr*.fasta.fai); do
    
    # extract the chromosome we are using:
    arrChr=(${INPUT//./ })
    
    # obtain the last coordinate of the inversion from the .fai file
    while read line; do 
        end=($(echo "$line" | awk '{print $2}'))
        start=($(echo "$line" | awk '{print $2-10001}'))
    done < "$INPUT"; 

    # use samtools 
    echo ${arrChr[2]}
    samtools faidx ${INPUT/.fai/} ${arrChr[2]}"_inv:"$start"-"$end > ${INPUT/.fasta.fai/.end.fasta}
    
done
~~~

Run Blast:

~~~bash
#!/bin/bash -l
 
ml load bioinfo-tools blast/2.11.0+

#place the files in an array:
array=(*[end,start].fasta)

#obtain the file
file="${array[$SLURM_ARRAY_TASK_ID]}"

#make blast database
makeblastdb -in ${file} -dbtype dna
~~~

Blast the end and beginning of CS3 and BS10 to the assemblies:

~~~bash
#!/bin/bash -l

# Load the modules:
ml load bioinfo-tools blast/2.11.0+

# Read the command line variable:
inversion=$chr

# place the files in an array:
fasta_files=(*.fasta)

# The list of individuals:
CS=("CS2" "CS4" "CS5" "CS7" "CS8" "CS10")
BS=("F2" "F3" "F4" "F5" "F6")

# obtain the file
file="${fasta_files[$SLURM_ARRAY_TASK_ID]}"

echo "Running: " ${file}

# conver the file name into an array
arrName=(${file//./ })
Name=${arrName[0]}


echo "Individual: " ${Name}

# Is the individual a CS individual?
if [[ " ${CS[*]} " =~ " ${Name} " ]]; then
    
    echo "This is a CS individual" ${Name}
    
    ref_start="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Inversion_scaffolds_and_assemblies/assemblies/inversions/CS10.hap1."$inversion".start.fasta"

    ref_end="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Inversion_scaffolds_and_assemblies/assemblies/inversions/CS10.hap1."$inversion".end.fasta"

    echo "Using the databases:"

    echo ${ref_start}

    echo ${ref_end}

    blastn -query ${file} -db ${ref_start} -evalue 1e-10 -outfmt 6 -max_target_seqs 25 -num_threads 1 -out ${file/.fasta/.blastn.start.result}

    blastn -query ${file} -db ${ref_end} -evalue 1e-10 -outfmt 6 -max_target_seqs 25 -num_threads 1 -out ${file/.fasta/.blastn.end.result}

else

    echo "Not a CS individual" ${Name}
    
    ref_start="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Inversion_scaffolds_and_assemblies/assemblies/inversions/BS3.hap2."$inversion".start.fasta"

    ref_end="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Inversion_scaffolds_and_assemblies/assemblies/inversions/BS3.hap2."$inversion".end.fasta"

    echo "Using the databases:"

    echo ${ref_start}

    echo ${ref_end}

    blastn -query ${file} -db ${ref_start} -evalue 1e-10 -outfmt 6 -max_target_seqs 25 -num_threads 1 -out ${file/.fasta/.blastn.start.result}

    blastn -query ${file} -db ${ref_end} -evalue 1e-10 -outfmt 6 -max_target_seqs 25 -num_threads 1 -out ${file/.fasta/.blastn.end.result}

fi
~~~

Use `find_breakpoints.py` to find the breakpoints from each blast result. The first version of the script will sort the output by qend, evalue and bitscore, and output the right most or left most coordinate from the best result.

The second version takes in the results for hap1 and hap2. In the first part of the script, we will sort the results by either send or sstart, and output the left or right coordinate for the result that includes the first and last coordinate of the 10kb piece of inversion. If, however, the end coordinate of the inversion is further left than the start coordinate, resulting in a inversion with a negative length, we sort the results of the "end" by bitscore instead and thake that coordinate. 

Sorting by send and sstart works well for chr6 and chr12, because sometimes the 10kb are split appart and make to two consecutive regions, and the longer piece will have a higher bitscore, but the shorter might be equaly good but not be the first entry in the table. This also works generally well for chromosome 17, except in four instances (F1 hap1, F3 hap1, F4 hap2 and F6 hap2), where I think the end of the inversion might be difficult to find (perhaps it's beyond the end of BS3 hap2). Then taking the best bitscore hit is fine, tough it doesn't fix the issue F3 hap1 or F6 hap2. For chr23, this approach fixes the negative lengths. Either way, in the end of the script, we choose the longest inversion from hap1 or hap2 to be output first. 

Of note that this approach results in more or less constant inversion lengths across individuals with a few exceptions. For example, F5 hap1 chr17 has a very long end at 29Mb, but looking at the blast result it realy does look like the coordinate is real because that is the best blast hit for the end of the inversion.  For CS2 hap2 chr23 the end is also long at 24Mb, but the other best blast hits are at similar coordinates. The same for F2 and CS7 hap2 in chr12. So, if we don't want to complicate this further, I think I will accept these results for now and advance on the TE annotation.

I think that for CS10 and BS3 we should always use the assembled version anyway for TE annotation.

The results are stored in the files `chr6_inversion_breakpoints.result.txt`, `chr12_inversion_breakpoints.result.txt`, `chr17_inversion_breakpoints.result.txt`, `chr23_inversion_breakpoints.result.txt`. 

~~~bash
for i in $(ls *.hap1.blastn.start.result); do python ~/Documents/Postdoc/Repositories/herring_selection/Inversion_Project/1.TE_annotation/find_breakpoints.py $i ${i/.hap1.blastn.start.result/.hap1.blastn.end.result} ${i/.hap1.blastn.start.result/.hap2.blastn.start.result} ${i/.hap1.blastn.start.result/.hap2.blastn.end.result}; done
~~~

Finally, extract the inversions from each assembly:

~~~bash
ml load bioinfo-tools samtools/1.12

for INPUT in $(ls *_breakpoints.result.txt); do
    
    # extract the chromosome we are using:
    arrChr=(${INPUT/_inversion_breakpoints.result.txt/})
    
    # obtain the information from the input file and run samtools
    while read line; do 

        #store the info into arrays (I think we could do this directly with the line above but oh well)
        individual=($(echo "$line" | awk '{print $1}'))
        haplotype=($(echo "$line" | awk '{print $2}'))
        start=($(echo "$line" | awk '{print $3}'))
        end=($(echo "$line" | awk '{print $4}'))

        #echo $individual
        #echo $haplotype
        #echo $start
        #echo $end

        echo "Input file: " ${individual}.${haplotype}.fasta 
        echo "Region: " ${arrChr[0]}_${haplotype}:$start-$end

        samtools faidx ${arrChr[0]}/${individual}.${haplotype}.fasta ${individual}_${haplotype}:$start-$end > ${arrChr[0]}/${individual}.${haplotype}.inversion.fasta
    
    done < "$INPUT"; 

done
~~~

## Run RepeatMasker

~~~bash
#!/bin/bash -l
 
# Load the modules:
module load bioinfo-tools RepeatMasker/4.1.0

# Read the command line variable:
inversion=$chr

# place the files in an array:
fasta_files=(*.inversion.fasta)

# obtain the file
file="${fasta_files[$SLURM_ARRAY_TASK_ID]}"

echo "Running: " ${file}

# conver the file name into an array
arrName=(${file//./ })
Individual=${arrName[0]}
Haplotype=${arrName[1]}

# Run RepeatMasker:
RepeatMasker -pa 1 -gff -no_is -a -dir ${Individual}_${Haplotype}_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta ${file}
~~~

We also run repeat masker for each genome (Not sure why I didn't use an array!)

~~~bash
module load bioinfo-tools RepeatMasker/4.1.0
ml load gnuparallel

RepeatMasker -pa 16 -gff -no_is -a -dir CS10_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS10.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS10_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS10.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS2_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS2.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS2_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS2.hap2.fa

RepeatMasker -pa 16 -gff -no_is -a -dir CS4_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS4.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS4_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS4.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS5_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS5.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS5_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS5.hap2.fa

RepeatMasker -pa 16 -gff -no_is -a -dir CS7_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS7.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS7_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS7.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS8_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS8.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir CS8_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta CS8.hap2.fa

RepeatMasker -pa 16 -gff -no_is -a -dir F1_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F1.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F1_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F1.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F2_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F2.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F2_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F2.hap2.fa

RepeatMasker -pa 16 -gff -no_is -a -dir F3_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F3.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F3_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F3.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F4_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F4.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F4_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F4.hap2.fa

RepeatMasker -pa 16 -gff -no_is -a -dir F5_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F5.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F5_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F5.hap2.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F6_hap1_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F6.hap1.fa
RepeatMasker -pa 16 -gff -no_is -a -dir F6_hap2_HerringCombinedDB -lib /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/TE_analysis/Hifiasm_assemblies/RepeatModeler/RepeatModelerParse/atlanticherring_combined_repDB_noprot_90.fasta F6.hap2.fa
~~~

Compile results. See excel file `inversion_coordinates.xlsx` 

~~~bash
for i in $(ls -d *HerringCombinedDB); do 

    o=${i/_HerringCombinedDB}
    
    input=$i/*tbl

    bases=($(grep "bases masked" $input | cut -d':' -f2 | cut -d' ' -f5))
    percent=($(grep "bases masked" $input | cut -d':' -f2 | cut -d' ' -f8))

    echo $o $bases $percent
done

~~~

## Plotting

You can find the code to plot the results in `plot_TE.R`.  The input and output files from this analyses can be found in [link]. This code was used to generate Figure 6C.