# dNdS analysis

This code assumes that the consensus genome sequences for each individual have already been produced. Check code in `genotype_call_pipeline.md` and `mapping_sprat.md` for how I did this for each individual and the outgroup.

### Modify gff file and select longest transcript per gene:

Changed the ":" in the ID field of the gff3 from ensembl by "_"

~~~bash
awk 'BEGIN{FS=OFS="\t"}; NR>32 {gsub(/\:/, "_", $9)} 1' Clupea_harengus.Ch_v2.0.2.104.gff3 > Clupea_harengus.Ch_v2.0.2.104.mod.gff3
~~~

Also, let's remove "biological functions"

~~~bash
awk 'BEGIN { OFS=FS="\t" } $3 !~ /^biological_region/' Clupea_harengus.Ch_v2.0.2.104.mod.gff3 > Clupea_harengus.Ch_v2.0.2.104.mod2.gff3
~~~

change with agat:

~~~bash
agat_convert_sp_gxf2gxf.pl -g Clupea_harengus.Ch_v2.0.2.104.mod2.gff3 -o Clupea_harengus.Ch_v2.0.2.104.mod2.agat.gff3
~~~

I noticed before that some genes are duplicated. This is because certain genes are duplicated in the herring genome, or they have the name external gene name. In the gff3 this field is called "Name", but the duplicated genes have their own ID from ENSEMBL, so they shouldn't come out as duplicated. agat has a tool to remove certain attributes, as we really only need ID and Parent in the gff file.

~~~bash
conda activate maker_annot_IsoSeq

agat_sp_manage_attributes.pl --gff Clupea_harengus.Ch_v2.0.2.104.mod2.agat.gff3 -p all --tag description,Name --out Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.gff3

#Let's try the suggestion of keeping longest transcript per gene and checking duplication levels:

agat_sp_keep_longest_isoform.pl --gff Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.gff3 --output Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.longest.gff3
~~~

I also created a list of transcripts where the sprat and the herring aligned with missing data bellow 20% and proportion of variable sites bellow than 0.2. See `mapping_sprat.md` to see how I generated `maff_alignments_version2.Miss20VariableSites20.genes.txt`

~~~bash
awk '{if($3=="mRNA") print $0}' /proj/snic2020-2-19/private/herring/users/mafalda/Herring_ISOSeq/Ensembl_gff3/Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.longest.gff3 | grep -f /proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/MKT_Sprat/maff_alignments_version2.Miss20VariableSites20.genes.txt | cut -f9 | cut -f1 -d';' | sed 's/ID=transcript_//' > ../scripts/transcript_list.txt
~~~

### Extract Coding sequences

Use agat to extract CDS sequences for each individual. Split sequences of each individual by transcript using seqkit. I am doing this just for the 35 individuals from the Baltic and Celtic Sea in our dataset.

~~~bash
#!/bin/bash -l

conda activate maker_annot_IsoSeq

# directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
INPUT_DIR="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Consensus"
GFF_DIR="/proj/snic2020-2-19/private/herring/users/mafalda/Herring_ISOSeq/Ensembl_gff3"
DIR_genomes=${WD}"/genomes"
DIR_transcripts=${WD}"/CDS_seqs"

#Individuals
Individuals=($(cat ${WD}/datasets/35individuals.txt))
Sample="${Individuals[$SLURM_ARRAY_TASK_ID]}"

# Generate a genome sequence
zcat ${INPUT_DIR}/chr*/${Sample}.chr*.modified.fa.gz > ${DIR_genomes}/${Sample}.modified.genome.fa 

# Extract agat:
agat_sp_extract_sequences.pl -g ${GFF_DIR}/Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.longest.gff3 -f ${DIR_genomes}/${Sample}.modified.genome.fa -t cds --output ${DIR_transcripts}/${Sample}.agat.transcript.cds.fasta

# Generate a genome sequence
mkdir ${DIR_transcripts}/${Sample}_transcripts_seqs

seqkit split ${DIR_transcripts}/${Sample}.agat.transcript.cds.fasta --by-id --out-dir ${DIR_transcripts}/${Sample}_transcripts_seqs --by-id-prefix ""

# Modify the header names and substitute by sample ID:

for fasta in $(ls ${DIR_transcripts}/${Sample}_transcripts_seqs/*.fasta); do echo $fasta; output_fasta=${fasta/.fasta/.mod.fasta}; awk -v ID=$Sample '/^>/ {printf ">%s\n", ID; next}{print $0}' $fasta > $output_fasta; done
~~~

Extract transcripts for sprat:

run_extract_CDS.sh
~~~bash
conda activate maker_annot_IsoSeq

# directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
INPUT_DIR="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/Consensus"
GFF_DIR="/proj/snic2020-2-19/private/herring/users/mafalda/Herring_ISOSeq/Ensembl_gff3"
DIR_genomes=${WD}"/genomes"
DIR_transcripts=${WD}"/CDS_seqs"

# Generate a genome sequence
zcat ${INPUT_DIR}/Sprat/EuSprat.chr*.modified.fa.gz > ${DIR_genomes}/EuSprat.modified.genome.fa 

# Extract agat:
agat_sp_extract_sequences.pl -g ${GFF_DIR}/Clupea_harengus.Ch_v2.0.2.104.mod2.agat.noName.longest.gff3 -f ${DIR_genomes}/EuSprat.modified.genome.fa -t cds --output ${DIR_transcripts}/EuSprat.agat.transcript.cds.fasta

# Generate a genome sequence
mkdir ${DIR_transcripts}/EuSprat_transcripts_seqs

seqkit split ${DIR_transcripts}/EuSprat.agat.transcript.cds.fasta --by-id --out-dir ${DIR_transcripts}/EuSprat_transcripts_seqs --by-id-prefix ""

# Modify the header names and substitute by sample ID:

for fasta in $(ls ${DIR_transcripts}/EuSprat_transcripts_seqs/*.fasta); do echo $fasta; output_fasta=${fasta/.fasta/.mod.fasta}; awk -v ID="EuSprat" '/^>/ {printf ">%s\n", ID; next}{print $0}' $fasta > $output_fasta; done
~~~

Combine each transcript of each individual into a single alignment using `transcript_list.txt`

run_generate_CDS_alignments.sh
~~~bash
#!/bin/bash -l
 
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
input_seqs=${WD}"/CDS_seqs"
output_seqs=${WD}"/CDS_seqs_by_transcript"


while read -r transcript; do 
    
    cat ${input_seqs}/*_transcripts_seqs/transcript_${transcript}.mod.fasta > ${output_seqs}/alignments/${transcript}.CDS.fasta

    /proj/snic2020-2-19/private/herring/users/mafalda/software/AMAS/amas/AMAS.py summary -i ${output_seqs}/alignments/${transcript}.CDS.fasta -f fasta -d dna -o ${output_seqs}/amas_summary/${transcript}.CDS.summary.txt

done < ${output_seqs}/transcript_list.txt
~~~

Copy the files with the transcripts of each inversion to the folder datasets, and transform them into new files with the names of the fasta alignments:

~~~bash
cp ../../../../Mutational_load/*transcripts.txt ./datasets/
cut -f2 -d"_" chr12_inversions_transcripts.txt | awk '{printf "%s.CDS.fasta\n", $1}' > chr12_inversions_transcripts_fastas.txt
cut -f2 -d"_" chr17_inversions_transcripts.txt | awk '{printf "%s.CDS.fasta\n", $1}' > chr17_inversions_transcripts_fastas.txt
cut -f2 -d"_" chr6_inversions_transcripts.txt | awk '{printf "%s.CDS.fasta\n", $1}' > chr6_inversions_transcripts_fastas.txt
cut -f2 -d"_" chr23_inversions_transcripts.txt | awk '{printf "%s.CDS.fasta\n", $1}' > chr23_inversions_transcripts_fastas.txt
~~~

Make a specific folder for each analysis

~~~bash
mkdir chr12_inversion; mkdir chr6_inversion; mkdir chr17_inversion; mkdir chr23_inversion; mkdir genome_wide
~~~

### Run dNdS analysis for each inversion and genome wide

The list of packages necessary to run `dnds.py` is listed in `mutation_conda.txt`

I wrote a custom script `dnds.py` that can be used to calculate dNdS by calling Bioconda.

#### Chr6 Inversion

run_chr6.sh
~~~bash 
#!/bin/bash -l

# Load conda
conda activate mutation

# Define directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
DATASETS=${WD}"/datasets"
INPUTS_ALGN=${WD}"/CDS_seqs_by_transcript/alignments"
# !!MODIFY!!
OUTPUT_RESULTS=${WD}"/dnds_calculations/chr6_inversion"

# Define input files !!MODIFY!!
FILE=$(cat ${DATASETS}/chr6_inversions_transcripts_fastas.txt | sed -n ${SLURM_ARRAY_TASK_ID}p)

# Define output files !!MODIFY!!
OUTFILE=${file/.CDS.fasta/.dnds.InvChr6.ML.txt}

# What are we analyzing?
echo ${FILE} ${OUTFILE}

# Enter the output directory
cd ${OUTPUT_RESULTS}

# Run python code: !!MODIFY!!
python3 -W ignore /proj/snic2020-2-19/private/herring/users/mafalda/software/dnds.py ${INPUTS_ALGN}/${FILE} ${DATASETS}/homozygotes_chr6_South.txt ${DATASETS}/homozygotes_chr6_North.txt ${OUTFILE}

~~~

#### Chr12 Inversion
run_chr12.sh
~~~bash 
#!/bin/bash -l

# Load conda
conda activate mutation

# Define directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
DATASETS=${WD}"/datasets"
INPUTS_ALGN=${WD}"/CDS_seqs_by_transcript/alignments"
# !!MODIFY!!
OUTPUT_RESULTS=${WD}"/dnds_calculations/chr12_inversion"

# Define input files !!MODIFY!!
FILE=$(cat ${DATASETS}/chr12_inversions_transcripts_fastas.txt | sed -n ${SLURM_ARRAY_TASK_ID}p)

# Define output files !!MODIFY!!
OUTFILE=${FILE/.CDS.fasta/.dnds.InvChr12.ML.txt}

# What are we analyzing?
echo ${FILE} ${OUTFILE}

# Enter the output directory
cd ${OUTPUT_RESULTS}

# Run python code: !!MODIFY!!
python3 -W ignore /proj/snic2020-2-19/private/herring/users/mafalda/software/dnds.py ${INPUTS_ALGN}/${FILE} ${DATASETS}/homozygotes_chr12_South.txt ${DATASETS}/homozygotes_chr12_North.txt ${OUTFILE}
~~~

#### Chr17 Inversion

run_chr17.sh
~~~bash 
#!/bin/bash -l
# Load conda
conda activate mutation

# Define directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
DATASETS=${WD}"/datasets"
INPUTS_ALGN=${WD}"/CDS_seqs_by_transcript/alignments"
# !!MODIFY!!
OUTPUT_RESULTS=${WD}"/dnds_calculations/chr17_inversion"

# Define input files !!MODIFY!!
FILE=$(cat ${DATASETS}/chr17_inversions_transcripts_fastas.txt | sed -n ${SLURM_ARRAY_TASK_ID}p)

# Define output files !!MODIFY!!
OUTFILE=${FILE/.CDS.fasta/.dnds.InvChr17.ML.txt}

# What are we analyzing?
echo ${FILE} ${OUTFILE}

# Enter the output directory
cd ${OUTPUT_RESULTS}

# Run python code: !!MODIFY!!
python3 -W ignore /proj/snic2020-2-19/private/herring/users/mafalda/software/dnds.py ${INPUTS_ALGN}/${FILE} ${DATASETS}/homozygotes_chr17_South.txt ${DATASETS}/homozygotes_chr17_North.txt ${OUTFILE}
~~~

#### Chr23 Inversion

run_chr23.sh
~~~bash 
#!/bin/bash -l

# Load conda
conda activate mutation

# Define directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
DATASETS=${WD}"/datasets"
INPUTS_ALGN=${WD}"/CDS_seqs_by_transcript/alignments"
# !!MODIFY!!
OUTPUT_RESULTS=${WD}"/dnds_calculations/chr23_inversion"

# Define input files !!MODIFY!!
FILE=$(cat ${DATASETS}/chr23_inversions_transcripts_fastas.txt | sed -n ${SLURM_ARRAY_TASK_ID}p)

# Define output files !!MODIFY!!
OUTFILE=${FILE/.CDS.fasta/.dnds.InvChr23.ML.txt}

# What are we analyzing?
echo ${FILE} ${OUTFILE}

# Enter the output directory
cd ${OUTPUT_RESULTS}

# Run python code: !!MODIFY!!
python3 -W ignore /proj/snic2020-2-19/private/herring/users/mafalda/software/dnds.py ${INPUTS_ALGN}/${FILE} ${DATASETS}/homozygotes_chr23_South.txt ${DATASETS}/homozygotes_chr23_North.txt ${OUTFILE}
~~~

#### Genome-wide dNdS

Create a list of all transcripts:

~~~bash
awk '{printf "/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS/CDS_seqs_by_transcript/alignments/%s.CDS.fasta\t%s.WG.txt\n", $1, $1}' ../CDS_seqs_by_transcript/transcript_list.txt > genome_wide/transcript_FILE_OUTFILE.txt
~~~

~~~bash
# Load conda
conda activate mutation

# Define directories
WD="/proj/snic2020-2-19/private/herring/users/mafalda/Inversion_project/GenomicsGenerals/F2_vcfs/Mutational_load_SFS/dNdS"
DATASETS=${WD}"/datasets"
INPUTS_ALGN=${WD}"/CDS_seqs_by_transcript/alignments"
# !!MODIFY!!
OUTPUT_RESULTS=${WD}"/dnds_calculations/genome_wide_background"

# Define output files !!MODIFY!!
OUTFILE=${file/.CDS.fasta/.dnds.GW.txt}

# What are we analyzing?
echo ${FILE} ${OUTFILE}

# Enter the output directory
cd ${OUTPUT_RESULTS}

# Run python code: !!MODIFY!!

while read -r input output; do python3 -W ignore /proj/snic2020-2-19/private/herring/users/mafalda/software/dnds.py $input ${DATASETS}/35individuals.txt ${DATASETS}/61AtlanticHerring.txt $output; done < transcript_FILE_OUTFILE.txt
~~~

## Plotting

Check R script `dNdS_plots.R`. Input and output files can be found at [link]